R/tidiers.R

```diff
@@ -0,0 +1,209 @@
+    coefs <- coef(x)
+        if (length(coefs) != nrow(ret)) {
+            coefs <- coefs |> tibble::enframe(name = "term", value = "estimate")
+            ret <- merge(coefs, ret, by = c("term", "estimate"))
+        }
+        if (conf.int) {
+            ci <- mmrm_confint_terms(x, level = conf.level)
+            ret <- merge(ret, ci, by = "term") |> tibble::as_tibble()
+        }
+        ret
+}
+
+#' @rdname tidiers
+#' @export
+#' @importFrom tibble as_tibble enframe
+glance.mmrm <- function (x) {
+  tibble::as_tibble(summary(x)$aic_list)
+}
+
+#' @rdname tidiers
+#' @export
+#' @importFrom tibble as_tibble
+augment.mmrm <- function(x, newdata = NULL,
+                         interval = c("none", "confidence", "prediction"),
+                         se_fit = interval!='none',
+                         type.residuals = c("response", "pearson", "normalized"), ...) {
+  .resid <- NULL
+  type.residuals <- match.arg(type.residuals, choices = c("response", "pearson", "normalized"))
+  if (missing(newdata)) {
+    newdata <- mmrm_getData(x)
+    .resid <- data.frame(.rownames = rownames(newdata), .resid = residuals(x, type = type.residuals) |> unname())
+  }
+  interval <- match.arg(interval, choices = c("none", "confidence", "prediction"))
+
+  ret <- mmrm_augment_newdata(x, newdata = newdata, .se_fit = se_fit, interval = interval)
+
+  if (!is.null(.resid)) {
+    ret <- merge(ret, .resid, by = '.rownames')
+    ret$.rownames <- as.numeric(ret$.rownames)
+    ret <- ret[order(ret$.rownames),]
+  }
+
+  ret |> tibble::as_tibble()
+}
+
+
+#' @importFrom tibble as_tibble
+mmrm_confint_terms <- function (x, ...)
+{
+    ci <- suppressMessages(confint(x, ...))
+    if (is.null(dim(ci))) {
+        ci <- matrix(ci, nrow = 1)
+        rownames(ci) <- names(coef(x))[1]
+    }
+    ci <- tibble::as_tibble(ci, rownames = "term", .name_repair = "minimal")
+    names(ci) <- c("term", "conf.low", "conf.high")
+    ci
+}
+
+
+mmrm_getData <- function(x){
+  get_all_vars(x, data = na.omit(x$data))
+}
+
+#' @importFrom tibble is_tibble
+has_rownames <- function (df) {
+  if (tibble::is_tibble(df)) {
+    return(FALSE)
+  }
+  any(rownames(df) != as.character(1:nrow(df)))
+}
+
+#' @importFrom tibble add_column
+as_augment_tibble <- function (data) {
+  if (inherits(data, "matrix") & is.null(colnames(data))) {
+    stop("The supplied `data`/`newdata` argument was an unnamed matrix. ",
+         "Please supply a matrix or dataframe with column names.")
+  }
+  tryCatch(df <- as_tibble(data), error = function(cnd) {
+    stop("Could not coerce data to `tibble`. Try explicitly passing a",
+         "dataset to either the `data` or `newdata` argument.",
+         call. = FALSE)
+  })
+  if (has_rownames(data)) {
+    df <- tibble::add_column(df, .rownames = rownames(data),
+                             .before = TRUE)
+  }
+  df
+}
+
+#' @importFrom rlang as_label f_lhs
+mmrm_augment_newdata <- function (x, data, newdata, .se_fit, interval = NULL, ...) {
+
```

Comment:
This is quite long and some repeated code, can you try DRY eg with helper functions?

Reply:
this was moved with slight changes from original broom script, will see how we can refactor to make it more efficient

Reply:
Function refactored, much simpler now